File: apple/826.most-profit-assigning-work.cpp
/*
 * @lc app=leetcode id=826 lang=cpp
 *
 * [826] Most Profit Assigning Work
 */
#include "bits/stdc++.h"

using namespace std;

// @lc code=start
class Solution {
public:
  int maxProfitAssignment(vector<int> &difficulty, vector<int> &profit,
                          vector<int> &worker) {
    vector<pair<int, int>> jobs;
    for (int i = 0; i < difficulty.size(); ++i) {
      jobs.push_back({difficulty[i], profit[i]});
    }
    sort(jobs.begin(), jobs.end());
    sort(worker.begin(), worker.end());

    int res = 0, i = 0, best = 0;

    for (const auto val : worker) {
      while (i < jobs.size() && val >= jobs[i].first) {
        best = max(best, jobs[i].second);
        ++i;
      }
      res += best;
    }

    return res;
  }
};
// @lc code=end

File: interview/apple/826.most-profit-assigning-work.py
#
# @lc app=leetcode id=826 lang=python3
#
# [826] Most Profit Assigning Work
#

# @lc code=start
from pprint import pprint
from typing import List
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        jobs = list(zip(difficulty, profit))
        jobs.sort()

        res = i = best =0
        n = len(jobs)
        
        for skill in sorted(worker):
            while i<n and skill >= jobs[i][0]:
                best = max(best, jobs[i][1]);
                i+=1
            res += best
        return res
        
# @lc code=end







    